### PR TITLE
Support custom metric for PyTorch

### PR DESCRIPTION
**Pull request by ForJadeForest**
add customized metric for orca

### CODE REVIEW

File: python/orca/example/learn/pytorch/brainMRI/brainMRI.py [hunk 1]
```diff
@@ -125,9 +127,7 @@ def model_creator(config):
 
 
 def optim_creator(model, config):
-    optimizer = optim.SGD(model.parameters(),
-                          lr=config.get("lr", 0.001),
-                          momentum=config.get("momentum", 0.9))
+    optimizer = optim.Adam(model.parameters(), lr=config.get("lr", 0.001), )
```
Comment (hkvision) [inline python/orca/example/learn/pytorch/brainMRI/brainMRI.py:130]: why modify this?

Comment (ForJadeForest) [reply python/orca/example/learn/pytorch/brainMRI/brainMRI.py:130]: The source codes use the Adam optimizer.

Comment (hkvision) [reply python/orca/example/learn/pytorch/brainMRI/brainMRI.py:130]: okay

File: python/orca/src/bigdl/orca/learn/metrics.py [hunk 2]
```diff
@@ -60,36 +61,65 @@ def get_name(self):
...
             else:
-                invalidInputError(False,
-                                  "Only orca metrics are supported, but get " +
-                                  metrics.__class__.__name__)
+                invalidInputError(False, "Only orca metrics and customized functions "
+                                         "are supported, but get " + m.__class__.__name__)
+        return metric_impls
+
+
+class MyMetric(Metric):
```
Comment (hkvision) [inline python/orca/src/bigdl/orca/learn/metrics.py:98]: rename to CustomizedMetric

File: python/orca/src/bigdl/orca/learn/metrics.py [hunk 3]
```diff
@@ -14,6 +14,7 @@
 # limitations under the License.
 #
 from abc import ABC, abstractmethod
+import types
```
Comment (hkvision) [inline python/orca/src/bigdl/orca/learn/metrics.py:17]: put this before line 16

File: python/orca/src/bigdl/orca/learn/metrics.py [hunk 4]
```diff
@@ -60,36 +61,65 @@ def get_name(self):
...
+        for m in metrics:
+            if isinstance(m, Metric):
                 metric_impls[m.get_name()] = m.get_metric(backend)
-            return metric_impls
-        else:
-            if isinstance(metrics, Metric):
-                return {metrics.get_name(): metrics.get_metric(backend)}
+            elif isinstance(m, types.FunctionType):
+                my_metric = MyMetric(m)
+                metric_impls[m.__name__] = my_metric.get_metric(backend)
```
Comment (hkvision) [inline python/orca/src/bigdl/orca/learn/metrics.py:91]: change to get_name?

File: python/orca/src/bigdl/orca/learn/metrics.py [hunk 5]
```diff
@@ -60,36 +61,65 @@ def get_name(self):
...
+                self.batch_metric_value = 0
+                self.compute_function = compute_function
+                self.step = 0
+
+            def __call__(self, preds, targets):
+                self.batch_metric_value += self.compute_function(preds, targets)
+                self.step += 1
+
+            def compute(self):
+                print(self.batch_metric_value, self.step)
```
Comment (hkvision) [inline python/orca/src/bigdl/orca/learn/metrics.py:116]: is this print necessary?

**Comment by hkvision** [comment]
1. Rebase with master to solve the conflicts
2. Add unit tests
3. Make the PR name more meaningful next time

File: python/orca/src/bigdl/orca/learn/metrics.py [hunk 6]
```diff
@@ -60,36 +61,66 @@ def get_name(self):
...
+
+        metric_impls = []
+        for m in metrics:
+            if isinstance(m, Metric):
+                metric_impls.append(m.get_metric(backend))
+            elif isinstance(m, types.FunctionType):
+                customized_metric = CustomizedMetric(m)
+                metric_impls.append(customized_metric.get_metric(backend))
+            else:
+                invalidInputError(False, "Only orca metrics and customized function "
```
Comment (hkvision) [inline python/orca/src/bigdl/orca/learn/metrics.py:75]: function -> functions

File: python/orca/src/bigdl/orca/learn/metrics.py [hunk 7]
```diff
@@ -60,36 +61,66 @@ def get_name(self):
...
-            if isinstance(metrics, Metric):
-                return {metrics.get_name(): metrics.get_metric(backend)}
+            elif isinstance(m, types.FunctionType):
+                my_metric = CustomizedMetric(m)
+                metric_impls[my_metric.get_name()] = my_metric.get_metric(backend)
             else:
-                invalidInputError(False,
-                                  "Only orca metrics are supported, but get " +
-                                  metrics.__class__.__name__)
+                invalidInputError(False, "Only orca metrics and customized function "
```
Comment (hkvision) [inline python/orca/src/bigdl/orca/learn/metrics.py:95]: same as above

File: python/orca/src/bigdl/orca/learn/metrics.py [hunk 8]
```diff
@@ -60,36 +61,66 @@ def get_name(self):
...
+
+
+class CustomizedMetric(Metric):
+    def __init__(self, compute_function):
+        self.compute = compute_function
+
+    def get_pytorch_metric(self):
+        from bigdl.orca.learn.pytorch.pytorch_metrics import PytorchMetric
+
+        class _Metirc(PytorchMetric):
```
Comment (hkvision) [inline python/orca/src/bigdl/orca/learn/metrics.py:107]: typo, should be Metric

File: python/orca/src/bigdl/orca/learn/metrics.py [hunk 9]
```diff
@@ -60,36 +61,66 @@ def get_name(self):
...
-                                  "Only orca metrics are supported, but get " +
-                                  metrics.__class__.__name__)
+                invalidInputError(False, "Only orca metrics and customized function "
+                                         "are supported, but get " + m.__class__.__name__)
+        return metric_impls
+
+
+class CustomizedMetric(Metric):
+    def __init__(self, compute_function):
+        self.compute = compute_function
```
Comment (hkvision) [inline python/orca/src/bigdl/orca/learn/metrics.py:102]: self.compute_function = func?

File: python/orca/test/bigdl/orca/learn/spark/test_estimator_for_bigdl.py [hunk 10]
```diff
@@ -296,7 +296,7 @@ def transform(df):
                                      model_dir=temp_dir_name,
                                      feature_preprocessing=SeqToTensor([2]),
                                      label_preprocessing=SeqToTensor([1]))
-            self.assertTrue('Only orca metrics are supported, but get str' in
+            self.assertTrue('Only orca metrics and customized function are supported, but get str' in
```
Comment (hkvision) [inline python/orca/test/bigdl/orca/learn/spark/test_estimator_for_bigdl.py:299]: functions

**Review by hkvision — APPROVED** [review]

**Comment by ForJadeForest** [comment]
http://10.112.231.51:18888/job/BigDL-Orca-PR-Validation/865/

The example test errors will be solved after merging.

**Pull request merged**